### DISCUSSION TITLE
Buttons with functions that return values?

### DISCUSSION THREAD

**Discussion by tataDan** [fyne-io/fyne#3166] [Q&A] [answered]
This code does not produce the desired result:

```
func getRandomImage() string {
   fmt.Println("Button clicked")
   return "Here is the random image URL"
}

func main() {
   ...
   x := ""
   randomImageBtn := widget.NewButton("Get Random Image", func() {
	x = getRandomImage()
   })

   fmt.Println(x)
   ...
}
```
The text "Button clicked" does appear in the terminal.  However x is not printed.

First of all, is anything like this possiblle in fyne?

If so, any suggestions as how to make this work would be appreciated.

**Comment by faide** [comment, marked as answer]
`fmt.Println(x)` is executed *before* your *callback* function is called (when the button is actually pressed by a user).

This is the reason you do not see anything. At the moment `fmt.Println(x)` is executed `x` is still equal to its initial value of `""`

**Comment by andydotxyz** [comment]
@faide is correct, the button callback happens when the user clicks it.
GUI based applications do not flow through main top to bottom then exit - their code executes based on when user interaction occurs.
For this reason callbacks will happen long after you define them - this is seen with buttons, dialogs and many other types of callback.
